Лабораторна робота №5
Завдання
Завантажте набір даних по медичному страхуванню за посиланням
За рахунок зміни гіперпараметрів нейронної мережі спробуйте досягнути найменшого значення функції втрат (mae).
Підготувати необхідні файли для захисту , за прикладом попередньої лабораторної роботи, але не викладати їх в репозиторій github до моменту захисту лабораторної роботи.
Викласти результати в репозиторій під час захисту лабораторної роботи.

In [4]:
import tensorflow as tf
import pandas as pd
from sklearn.model_selection import train_test_split
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

# Завантаження даних
data = pd.read_csv('medical_insurance.csv')

# Преобразування категорійних змінних в числа
data_one_hot = pd.get_dummies(data, dtype=int)

# Створення X та y значень
X = data_one_hot.drop("charges", axis=1)
y = data_one_hot["charges"]

# Розділення на навчальний і тестовий набори
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=13)

# Створення моделі
model = Sequential([
    Dense(units=64, activation='relu', input_shape=(X.shape[1],)),
    Dense(units=32, activation='relu'),
    Dense(units=1)
])

# Компіляція та тренування моделі
model.compile(optimizer='adam', loss='mae', metrics=['mae'])
model.fit(X_train, y_train, epochs=100, validation_data=(X_test, y_test))

# Оцінка моделі на тестовому наборі
model.evaluate(X_test, y_test)


Epoch 1/100
34/34 [==============================] - 1s 10ms/step - loss: 13454.7061 - mae: 13454.7061 - val_loss: 12406.6035 - val_mae: 12406.6035
Epoch 2/100
34/34 [==============================] - 0s 4ms/step - loss: 13411.6357 - mae: 13411.6357 - val_loss: 12352.2188 - val_mae: 12352.2188
Epoch 3/100
34/34 [==============================] - 0s 4ms/step - loss: 13334.0068 - mae: 13334.0068 - val_loss: 12246.3291 - val_mae: 12246.3291
Epoch 4/100
34/34 [==============================] - 0s 4ms/step - loss: 13185.8779 - mae: 13185.8779 - val_loss: 12049.8564 - val_mae: 12049.8564
Epoch 5/100
34/34 [==============================] - 0s 4ms/step - loss: 12921.7949 - mae: 12921.7949 - val_loss: 11713.8037 - val_mae: 11713.8037
Epoch 6/100
34/34 [==============================] - 0s 4ms/step - loss: 12480.6113 - mae: 12480.6113 - val_loss: 11160.9355 - val_mae: 11160.9355
Epoch 7/100
34/34 [==============================] - 0s 5ms/step - loss: 11799.9736 - mae: 11799.9736 - val_loss: 103

[5537.693359375, 5537.693359375]

In [5]:
import tensorflow as tf
import pandas as pd
from sklearn.model_selection import train_test_split
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

# Завантаження даних
data = pd.read_csv('medical_insurance.csv')

# Преобразування категорійних змінних в числа
data_one_hot = pd.get_dummies(data, dtype=int)

# Створення X та y значень
X = data_one_hot.drop("charges", axis=1)
y = data_one_hot["charges"]

# Розділення на навчальний і тестовий набори
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=13)

# Створення моделі з додатковим шаром та більшою кількістю нейронів
model = Sequential([
    Dense(units=64, activation='relu', input_shape=(X.shape[1],)),
    Dense(units=128, activation='relu'),  # Новий шар із більшою кількістю нейронів
    Dense(units=32, activation='relu'),
    Dense(units=1)
])

# Компіляція та тренування моделі
model.compile(optimizer='adam', loss='mae', metrics=['mae'])
model.fit(X_train, y_train, epochs=100, validation_data=(X_test, y_test))

# Оцінка моделі на тестовому наборі
model.evaluate(X_test, y_test)


Epoch 1/100
34/34 [==============================] - 2s 16ms/step - loss: 13437.0596 - mae: 13437.0596 - val_loss: 12347.8066 - val_mae: 12347.8066
Epoch 2/100
34/34 [==============================] - 0s 4ms/step - loss: 13200.6572 - mae: 13200.6572 - val_loss: 11850.5098 - val_mae: 11850.5098
Epoch 3/100
34/34 [==============================] - 0s 4ms/step - loss: 12074.1953 - mae: 12074.1953 - val_loss: 9777.4990 - val_mae: 9777.4990
Epoch 4/100
34/34 [==============================] - 0s 5ms/step - loss: 9114.8174 - mae: 9114.8174 - val_loss: 7070.8760 - val_mae: 7070.8760
Epoch 5/100
34/34 [==============================] - 0s 5ms/step - loss: 7604.7715 - mae: 7604.7715 - val_loss: 7012.5757 - val_mae: 7012.5757
Epoch 6/100
34/34 [==============================] - 0s 5ms/step - loss: 7522.4624 - mae: 7522.4624 - val_loss: 6969.6694 - val_mae: 6969.6694
Epoch 7/100
34/34 [==============================] - 0s 4ms/step - loss: 7470.7500 - mae: 7470.7500 - val_loss: 6894.7954 - val_mae

[2309.803955078125, 2309.803955078125]

Зменшення значення функції втрат (MAE) з 5537.69 до 2309.80 свідчить про покращення моделі після внесення змін у гіперпараметри. 

In [6]:
import tensorflow as tf
import pandas as pd
from sklearn.model_selection import train_test_split
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

# Завантаження даних
data = pd.read_csv('medical_insurance.csv')

# Преобразування категорійних змінних в числа
data_one_hot = pd.get_dummies(data, dtype=int)

# Створення X та y значень
X = data_one_hot.drop("charges", axis=1)
y = data_one_hot["charges"]

# Розділення на навчальний і тестовий набори
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=13)

# Створення моделі з більшою кількістю шарів і нейронів
model = Sequential([
    Dense(units=64, activation='relu', input_shape=(X.shape[1],)),
    Dense(units=128, activation='relu'),  
    Dense(units=256, activation='relu'),  # Доданий шар із більшою кількістю нейронів
    Dense(units=64, activation='relu'),   # Доданий шар
    Dense(units=32, activation='relu'),
    Dense(units=1)
])

# Компіляція та тренування моделі з більшими параметрами
model.compile(optimizer='adam', loss='mae', metrics=['mae'])
model.fit(X_train, y_train, epochs=150, validation_data=(X_test, y_test))  # Збільшено кількість епох

# Оцінка моделі на тестовому наборі
model.evaluate(X_test, y_test)


Epoch 1/150
34/34 [==============================] - 2s 13ms/step - loss: 13214.2314 - mae: 13214.2314 - val_loss: 11151.6113 - val_mae: 11151.6113
Epoch 2/150
34/34 [==============================] - 0s 5ms/step - loss: 8979.3887 - mae: 8979.3887 - val_loss: 6952.2686 - val_mae: 6952.2686
Epoch 3/150
34/34 [==============================] - 0s 5ms/step - loss: 7459.0298 - mae: 7459.0298 - val_loss: 6786.5400 - val_mae: 6786.5400
Epoch 4/150
34/34 [==============================] - 0s 5ms/step - loss: 7291.1167 - mae: 7291.1167 - val_loss: 6620.2148 - val_mae: 6620.2148
Epoch 5/150
34/34 [==============================] - 0s 5ms/step - loss: 7145.2529 - mae: 7145.2529 - val_loss: 6441.2500 - val_mae: 6441.2500
Epoch 6/150
34/34 [==============================] - 0s 6ms/step - loss: 6992.7466 - mae: 6992.7466 - val_loss: 6316.9497 - val_mae: 6316.9497
Epoch 7/150
34/34 [==============================] - 0s 5ms/step - loss: 6856.6069 - mae: 6856.6069 - val_loss: 6144.8496 - val_mae: 6144

[1439.88720703125, 1439.88720703125]

Зменшення значення функції втрат (MAE) з 2309.80 до 1439.89 свідчить про подальше покращення моделі після зміни гіперпараметрів.

In [7]:
import tensorflow as tf
import pandas as pd
from sklearn.model_selection import train_test_split
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

# Завантаження даних
data = pd.read_csv('medical_insurance.csv')

# Преобразування категорійних змінних в числа
data_one_hot = pd.get_dummies(data, dtype=int)

# Створення X та y значень
X = data_one_hot.drop("charges", axis=1)
y = data_one_hot["charges"]

# Розділення на навчальний і тестовий набори
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=13)

# Створення моделі з більшою кількістю шарів і нейронів
model = Sequential([
    Dense(units=128, activation='relu', input_shape=(X.shape[1],)),
    Dense(units=256, activation='relu'),  
    Dense(units=512, activation='relu'),  # Доданий шар із більшою кількістю нейронів
    Dense(units=256, activation='relu'),   # Доданий шар
    Dense(units=128, activation='relu'),
    Dense(units=64, activation='relu'),
    Dense(units=32, activation='relu'),
    Dense(units=1)
])

# Компіляція та тренування моделі з більшими параметрами
model.compile(optimizer='adam', loss='mae', metrics=['mae'])
model.fit(X_train, y_train, epochs=200, validation_data=(X_test, y_test))  # Збільшено кількість епох

# Оцінка моделі на тестовому наборі
model.evaluate(X_test, y_test)


Epoch 1/200
34/34 [==============================] - 3s 16ms/step - loss: 12076.4766 - mae: 12076.4766 - val_loss: 8678.4697 - val_mae: 8678.4697
Epoch 2/200
34/34 [==============================] - 0s 10ms/step - loss: 7751.9814 - mae: 7751.9814 - val_loss: 6886.2876 - val_mae: 6886.2876
Epoch 3/200
34/34 [==============================] - 0s 10ms/step - loss: 7281.6348 - mae: 7281.6348 - val_loss: 6762.0239 - val_mae: 6762.0239
Epoch 4/200
34/34 [==============================] - 0s 8ms/step - loss: 7066.8066 - mae: 7066.8066 - val_loss: 6408.0093 - val_mae: 6408.0093
Epoch 5/200
34/34 [==============================] - 0s 7ms/step - loss: 6816.7046 - mae: 6816.7046 - val_loss: 6107.6221 - val_mae: 6107.6221
Epoch 6/200
34/34 [==============================] - 0s 8ms/step - loss: 6777.5317 - mae: 6777.5317 - val_loss: 6116.4727 - val_mae: 6116.4727
Epoch 7/200
34/34 [==============================] - 0s 8ms/step - loss: 6666.3047 - mae: 6666.3047 - val_loss: 6012.1650 - val_mae: 6012

[1485.0947265625, 1485.0947265625]